server2/Arcadia.Assistant/Arcadia.Assistant.ExternalStorages.SharepointOnline/SharepointStorage.cs

```diff
@@ -0,0 +1,170 @@
+﻿namespace Arcadia.Assistant.ExternalStorages.SharepointOnline
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Net.Http;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    using Abstractions;
+
+    using Contracts;
+
+    using SharepointApiModels;
+
+    public class SharepointStorage : IExternalStorage
+    {
+        private readonly ISharepointConditionsCompiler conditionsCompiler;
+        private readonly ISharepointOnlineConfiguration configuration;
+        private readonly ISharepointRequestExecutor requestExecutor;
+
+        public SharepointStorage(
+            ISharepointOnlineConfiguration configuration,
+            ISharepointRequestExecutor requestExecutor,
+            ISharepointFieldsMapper sharePointFieldsMapper,
+            ISharepointConditionsCompiler sharePointConditionsCompiler)
+        {
+            this.configuration = configuration;
+            this.requestExecutor = requestExecutor;
+            this.FieldsMapper = sharePointFieldsMapper;
+            this.conditionsCompiler = sharePointConditionsCompiler;
+            this.StorageItemJsonConvertor = new StorageItemJsonConvertor(this.FieldsMapper);
+        }
+
+        private ISharepointFieldsMapper FieldsMapper { get; }
+
+        private StorageItemJsonConvertor StorageItemJsonConvertor { get; }
+
+        public async Task<IEnumerable<StorageItem>> GetItems(string list, IEnumerable<ICondition>? conditions, CancellationToken cancellationToken)
+        {
+            var listItems = await this.GetListItems(list, conditions, cancellationToken);
+
+            return listItems?.ToArray() ?? new StorageItem[0];
+        }
+
+        public async Task<StorageItem> AddItem(string list, StorageItem item, CancellationToken cancellationToken)
+        {
+            var listItemType = await this.GetListItemType(list, cancellationToken);
+            var requestData = this.StorageItemJsonConvertor.StorageItemToRequestJson(item, listItemType);
+
+            var request = SharepointRequest
+                .Create(HttpMethod.Post, this.GetListItemsUrl(list, false))
+                .WithAcceptHeader("application/json;odata=verbose")
+                .WithContentString(requestData);
+
+            return await this.requestExecutor.ExecuteSharepointRequest<StorageItem>(request, cancellationToken);
+        }
+
+        public async Task UpdateItem(string list, StorageItem item, CancellationToken cancellationToken)
+        {
+            var idConditions = new[] { new EqualCondition(x => x.Id, item.Id) };
+            var listItems = (await this.GetListItems(list, idConditions, cancellationToken)).ToArray();
+
+            this.EnsureSingleItemReturned(listItems);
+
+            var existingListItem = listItems.First();
+
+            var updateItemUrl = this.GetListItemsUrl(list, false);
```

Comment:
`$"{this.GetListItemsUrl(list, false)}({existingListItem.Id})"`

Reply:
Done.